go/vt/vttablet/tabletserver/stream_consolidator.go

```diff
@@ -0,0 +1,319 @@
+/*
+Copyright 2021 The Vitess Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package tabletserver
+
+import (
+	"sync"
+	"sync/atomic"
+
+	"vitess.io/vitess/go/sqltypes"
+	vtrpcpb "vitess.io/vitess/go/vt/proto/vtrpc"
+	"vitess.io/vitess/go/vt/vterrors"
+	"vitess.io/vitess/go/vt/vttablet/tabletserver/tabletenv"
+)
+
+const streamBufferSize = 8
+
+// StreamConsolidator is a data structure capable of merging several identical streaming queries so only
+// one query is executed in MySQL and its response is fanned out to all the clients simultaneously.
+type StreamConsolidator struct {
+	mu                             sync.Mutex
+	inflight                       map[string]*streamInFlight
+	memory                         int64
+	maxMemoryTotal, maxMemoryQuery int64
+	blocking                       bool
+	cleanup                        StreamCallback
+}
+
+// NewStreamConsolidator allocates a stream consolidator. The consolidator will use up to maxMemoryTotal
+// bytes in order to allow simultaneous queries to "catch up" to each other. Each individual stream will
+// only use up to maxMemoryQuery bytes of memory as a history buffer to catch up.
+func NewStreamConsolidator(maxMemoryTotal, maxMemoryQuery int64, cleanup StreamCallback) *StreamConsolidator {
+	return &StreamConsolidator{
+		inflight:       make(map[string]*streamInFlight),
+		maxMemoryTotal: maxMemoryTotal,
+		maxMemoryQuery: maxMemoryQuery,
+		blocking:       false,
+		cleanup:        cleanup,
+	}
+}
+
+// StreamCallback is a function that yields every Result object from a streaming query
```

Comment:
I don't understand this comment. This function will be called with the result objects, it won't yield anything, right?